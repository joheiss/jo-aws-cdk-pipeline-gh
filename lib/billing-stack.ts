import { Stack, StackProps } from 'aws-cdk-lib';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { Construct } from 'constructs';
import { CostBudgetWarning } from './constructs/budget-construct';

export class BillingStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    // create SNS topic for budget warning
    const topic =  new Topic(this, 'MonthlyCostBudgetWarningTopic', {
      displayName: 'Monthly Cost Budget Warning Topic',
      topicName: 'MonthlyCostBudgetWarningTopic',
  });

    // create budget warning 
    const budgetWarning = new CostBudgetWarning(this, 'Budget-Warning', {
      name: 'MonthlyCostBudgetWarning',
      limitAmount: 10,
      snsTopicArn: topic.topicArn,
    });

  }
}
