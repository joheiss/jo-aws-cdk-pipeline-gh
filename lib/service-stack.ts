import { HttpApi } from "@aws-cdk/aws-apigatewayv2-alpha";
import { HttpLambdaIntegration } from "@aws-cdk/aws-apigatewayv2-integrations-alpha";
import { CfnOutput, Duration, Stack, StackProps } from "aws-cdk-lib";
import { Statistic, TreatMissingData } from "aws-cdk-lib/aws-cloudwatch";
import {
  LambdaDeploymentConfig,
  LambdaDeploymentGroup,
} from "aws-cdk-lib/aws-codedeploy";
import {
  Alias,
  Architecture,
  CfnParametersCode,
  Code,
  Function,
  Runtime,
} from "aws-cdk-lib/aws-lambda";
import { Topic } from "aws-cdk-lib/aws-sns";
import { EmailSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import { Construct } from "constructs";
import { ServiceHealthCanary } from "./constructs/service-health-canary";

interface ServiceStackProps extends StackProps {
  stageName: string;
}
export class ServiceStack extends Stack {
  public readonly serviceCode: CfnParametersCode;
  public readonly serviceEndpointOutput: CfnOutput;
  private readonly alias: Alias;
  private readonly api: HttpApi;

  constructor(scope: Construct, id: string, props: ServiceStackProps) {
    super(scope, id, props);

    this.serviceCode = Code.fromCfnParameters();

    // create lambda function - typescript
    const lambdaFn = new Function(this, "LambdaFunction", {
      architecture: Architecture.ARM_64,
      memorySize: 128,
      timeout: Duration.seconds(3),
      runtime: Runtime.NODEJS_14_X,
      handler: "src/lambda.handler",
      code: this.serviceCode,
      functionName: `ServiceLambda${props.stageName}`,
      description: `generated on ${new Date().toISOString()}`, // make sure a change is re-deployed on every CDK build
    });

    // create a lambda alias
    this.alias = new Alias(this, "ServiceLambdaAlias", {
      version: lambdaFn.currentVersion,
      aliasName: `ServiceLambdaAlias${props.stageName}`,
    });

    // create HTTP API
    this.api = new HttpApi(this, "ServiceApi", {
      defaultIntegration: new HttpLambdaIntegration(
        "LambdaIntegration",
        this.alias
      ),
      apiName: `JoService${props.stageName}`,
    });

    // create Lambda Deployment Group - for Prod stage only
    if (props.stageName === "Prod") {
      this._createLambdaDeploymentGroup(props.stageName);
      const alarmTopic = new Topic(this, "ServiceAlarmTopic", {
        topicName: "ServiceAlarmTopic",
      });
      alarmTopic.addSubscription(new EmailSubscription("aws-lab@jovisco.de"));
      this._createServiceHealthCanary(alarmTopic);
    }

    // create Output for API endpoint
    this.serviceEndpointOutput = new CfnOutput(this, "ApiEndPointOutput", {
      exportName: `JoServiceEndpoint${props.stageName}`,
      value: this.api.apiEndpoint,
      description: "API Endpoint",
    });
  }

  private _createLambdaDeploymentGroup(stageName: string): void {
    new LambdaDeploymentGroup(this, "DeploymentGroup", {
      alias: this.alias,
      deploymentConfig: LambdaDeploymentConfig.CANARY_10PERCENT_5MINUTES,
      autoRollback: {
        deploymentInAlarm: true,
      },
      alarms: [
        this.api
          .metricServerError()
          .with({
            period: Duration.minutes(1),
            statistic: Statistic.SUM,
          })
          .createAlarm(this, "ServiceErrorAlarm", {
            threshold: 1,
            alarmDescription: "Service is experiencing errors",
            alarmName: `ServiceErrorAlarm${stageName}`,
            evaluationPeriods: 1,
            treatMissingData: TreatMissingData.NOT_BREACHING,
          }),
      ],
    });
  }

  private _createServiceHealthCanary(alarmTopic: Topic): void {
    new ServiceHealthCanary(this, "ServiceHealthCanary", {
      apiEndpoint: this.api.apiEndpoint,
      canaryName: "service-canary",
      alarmTopic,
    });
  }
}
