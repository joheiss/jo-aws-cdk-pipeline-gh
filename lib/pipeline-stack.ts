import { SecretValue, Stack, StackProps } from "aws-cdk-lib";
import {
  BuildEnvironmentVariableType,
  BuildSpec,
  LinuxBuildImage,
  PipelineProject,
} from "aws-cdk-lib/aws-codebuild";
import { Artifact, IStage, Pipeline } from "aws-cdk-lib/aws-codepipeline";
import {
  CloudFormationCreateUpdateStackAction,
  CodeBuildAction,
  CodeBuildActionType,
  GitHubSourceAction,
} from "aws-cdk-lib/aws-codepipeline-actions";
import { Construct } from "constructs";
import { ServiceStack } from "./service-stack";
import { BillingStack } from "./billing-stack";
import { Topic } from "aws-cdk-lib/aws-sns";
import { EmailSubscription } from "aws-cdk-lib/aws-sns-subscriptions";
import { SnsTopic } from "aws-cdk-lib/aws-events-targets";
import { EventField, RuleTargetInput } from "aws-cdk-lib/aws-events";

interface PipelineStackProps extends StackProps {
  ownerEmail?: string;
}

export class PipelineStack extends Stack {
  private readonly pipeline: Pipeline;
  private cdkSourceOutput: Artifact;
  private serviceSourceOutput: Artifact;
  private cdkBuildOutput: Artifact;
  private serviceBuildOutput: Artifact;
  private readonly snsTopic: Topic;

  constructor(scope: Construct, id: string, props?: PipelineStackProps) {
    super(scope, id, props);

    if (props?.ownerEmail) {
      this.snsTopic = this._createSnsTopic(props.ownerEmail);
    }
    this.pipeline = this._createCodePipeline();
  }

  public addDeployStage(serviceStack: ServiceStack, stageName: string): IStage {
    return this.pipeline.addStage({
      stageName,
      actions: [
        new CloudFormationCreateUpdateStackAction({
          actionName: "Service_Update",
          stackName: serviceStack.stackName,
          templatePath: this.cdkBuildOutput.atPath(
            `${serviceStack.stackName}.template.json`
          ),
          adminPermissions: true,
          parameterOverrides: {
            ...serviceStack.serviceCode.assign(
              this.serviceBuildOutput.s3Location
            ),
          },
          extraInputs: [this.serviceBuildOutput],
        }),
      ],
    });
  }

  public addTestToStage(stage: IStage, serviceEndpoint: string): void {
    const testAction = new CodeBuildAction({
      actionName: "Service_Test",
      input: this.serviceSourceOutput,
      project: new PipelineProject(this, "ServiceTestBuildProject", {
        environment: {
          buildImage: LinuxBuildImage.STANDARD_5_0,
        },
        buildSpec: BuildSpec.fromSourceFilename(
          "buildspecs/test-buildspec.yml"
        ),
      }),
      environmentVariables: {
        SERVICE_ENDPOINT: {
          value: serviceEndpoint,
          type: BuildEnvironmentVariableType.PLAINTEXT,
        },
      },
      type: CodeBuildActionType.TEST,
      runOrder: 2,
    });
    stage.addAction(testAction);
    testAction.onStateChange(
      "TestFailed",
      new SnsTopic(this.snsTopic, {
        message: RuleTargetInput.fromText(
          `Test failed. See details here: ${EventField.fromPath(
            "$.detail.execution-result.external-execution-url"
          )}`
        ),
      }),
      {
        ruleName: "TestFailed",
        eventPattern: {
          detail: {
            state: ["FAILED"],
          },
        },
        description: "Test has failed",
      }
    );
  }

  public addBillingStackToDeployStage(
    billingStack: BillingStack,
    stage: IStage
  ): void {
    stage.addAction(
      new CloudFormationCreateUpdateStackAction({
        actionName: "BudgetWarning_Update",
        stackName: billingStack.stackName,
        templatePath: this.cdkBuildOutput.atPath(
          `${billingStack.stackName}.template.json`
        ),
        adminPermissions: true,
      })
    );
  }

  private _createSnsTopic(ownerEmail: string): Topic {
    const topic = new Topic(this, "NotifyOnFailedPipeline", {
      topicName: "NotifyOnFailedPipeline",
    });

    topic.addSubscription(new EmailSubscription(ownerEmail));

    return topic;
  }

  private _createCodePipeline(): Pipeline {
    // create code pipeline
    const pipeline = new Pipeline(this, "Pipeline", {
      pipelineName: "Pipeline",
      crossAccountKeys: false,
      restartExecutionOnUpdate: true,
    });

    // add source stage
    this._addSourceStage(pipeline);

    // add build stage
    this._addBuildStage(pipeline);

    // add update stage
    this._addUpdateStage(pipeline);

    return pipeline;
  }

  private _addSourceStage(pipeline: Pipeline): void {
    // create source output artifact
    this.cdkSourceOutput = new Artifact("CdkSourceOutput");
    this.serviceSourceOutput = new Artifact("ServiceSourceOutput");

    // add source stage
    pipeline.addStage({
      stageName: "Source",
      actions: [
        new GitHubSourceAction({
          actionName: "Pipeline_Source",
          owner: "joheiss",
          repo: "jo-aws-cdk-pipeline-gh",
          branch: "main",
          oauthToken: SecretValue.secretsManager("github-token"),
          output: this.cdkSourceOutput,
        }),
        new GitHubSourceAction({
          actionName: "Service_Source",
          owner: "joheiss",
          repo: "jo-aws-express-app",
          branch: "main",
          oauthToken: SecretValue.secretsManager("github-token"),
          output: this.serviceSourceOutput,
        }),
      ],
    });
  }

  private _addBuildStage(pipeline: Pipeline): void {
    // create build output artifact
    this.cdkBuildOutput = new Artifact("CdkBuildOutput");
    this.serviceBuildOutput = new Artifact("ServiceBuildOutput");

    // add build stage
    pipeline.addStage({
      stageName: "Build",
      actions: [
        new CodeBuildAction({
          actionName: "Pipeline_Build",
          input: this.cdkSourceOutput,
          outputs: [this.cdkBuildOutput],
          project: new PipelineProject(this, "CdkBuildProject", {
            environment: {
              buildImage: LinuxBuildImage.STANDARD_5_0,
            },
            buildSpec: BuildSpec.fromSourceFilename(
              "buildspecs/cdk-buildspec.yml"
            ),
          }),
        }),
        new CodeBuildAction({
          actionName: "Service_Build",
          input: this.serviceSourceOutput,
          outputs: [this.serviceBuildOutput],
          project: new PipelineProject(this, "ServiceBuildProject", {
            environment: {
              buildImage: LinuxBuildImage.STANDARD_5_0,
            },
            buildSpec: BuildSpec.fromSourceFilename(
              "buildspecs/service-buildspec.yml"
            ),
          }),
        }),
      ],
    });
  }

  private _addUpdateStage(pipeline: Pipeline): void {
    pipeline.addStage({
      stageName: "Update",
      actions: [
        new CloudFormationCreateUpdateStackAction({
          actionName: "Pipeline_Update",
          stackName: "PipelineStack",
          templatePath: this.cdkBuildOutput.atPath(
            "PipelineStack.template.json"
          ),
          adminPermissions: true,
        }),
      ],
    });
  }
}
