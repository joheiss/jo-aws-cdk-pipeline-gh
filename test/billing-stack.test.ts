import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { BillingStack } from "../lib/billing-stack";

describe('Monthly Cost Budget Warning Stack', () => {

    let app: App;
    let stack: BillingStack;
    
    beforeEach(() => {
        app = new App();
        stack = new BillingStack(app, 'MontlyCostBudgetWarningStack');
    });

    test('Stack', () => {
        const cfn = Template.fromStack(stack).toJSON();
        expect(cfn).toMatchSnapshot();
    });
});