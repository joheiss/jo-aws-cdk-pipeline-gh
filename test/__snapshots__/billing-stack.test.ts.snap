// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Monthly Cost Budget Warning Stack Stack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "BudgetWarningmyBudget7091B012": Object {
      "Properties": Object {
        "Budget": Object {
          "BudgetLimit": Object {
            "Amount": 10,
            "Unit": "USD",
          },
          "BudgetName": "MonthlyCostBudgetWarning",
          "BudgetType": "COST",
          "TimeUnit": "MONTHLY",
        },
        "NotificationsWithSubscribers": Array [
          Object {
            "Notification": Object {
              "ComparisonOperator": "GREATER_THAN",
              "NotificationType": "ACTUAL",
              "Threshold": 100,
              "ThresholdType": "PERCENTAGE",
            },
            "Subscribers": Array [
              Object {
                "Address": Object {
                  "Ref": "MonthlyCostBudgetWarningTopic938BE3FB",
                },
                "SubscriptionType": "SNS",
              },
            ],
          },
        ],
      },
      "Type": "AWS::Budgets::Budget",
    },
    "MonthlyCostBudgetWarningTopic938BE3FB": Object {
      "Properties": Object {
        "DisplayName": "Monthly Cost Budget Warning Topic",
        "TopicName": "MonthlyCostBudgetWarningTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
